adapters/repos/db/vector/hnsw/commit_logger_snapshot.go

```diff
@@ -0,0 +1,1008 @@
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/compressionhelpers"
+	entcfg "github.com/weaviate/weaviate/entities/config"
+	"github.com/weaviate/weaviate/entities/diskio"
+	enterrors "github.com/weaviate/weaviate/entities/errors"
+)
+
+const checkpointChunkSize = 100_000
+
+const (
+	SnapshotCompressionTypePQ = iota + 1
+	SnapshotCompressionTypeSQ
+)
+
+// feature flag to disable snapshots
+func snapshotsDisabled() bool {
+	return entcfg.Enabled(os.Getenv("DISABLE_SNAPSHOTS"))
+}
+
+func snapshotTimestamp(path string) (int64, error) {
+	return asTimeStamp(strings.TrimSuffix(filepath.Base(path), ".snapshot"))
+}
+
+func snapshotDirectory(rootPath, name string) string {
+	return fmt.Sprintf("%s/%s.hnsw.snapshot.d", rootPath, name)
+}
+
+// Creates a snapshot of the commit log and returns the deserialized state.
+// The snapshot is created from the last snapshot if any, or from the entire commit
+// log.
+// The snapshot state stops at the last commit log file that satisfies these conditions:
+// - the file is condensed
+// - the file cannot be combined further with the next file
+// - the file is not the last condensed commit log file
+// These conditions ensure immutability of the files used to create the snapshot.
+func (l *hnswCommitLogger) CreateSnapshot() (*DeserializationResult, int64, error) {
+	return l.createOrLoadSnapshot(false)
+}
+
+// CreateOrLoadSnapshot works like CreateSnapshot, but it will always load the
+// last snapshot. It is used at startup to automatically create a snapshot
+// while loading the commit log, to avoid having to load the commit log again.
+func (l *hnswCommitLogger) CreateOrLoadSnapshot() (*DeserializationResult, int64, error) {
+	return l.createOrLoadSnapshot(true)
+}
+
+func (l *hnswCommitLogger) createOrLoadSnapshot(load bool) (*DeserializationResult, int64, error) {
+	err := os.MkdirAll(snapshotDirectory(l.rootPath, l.id), 0o755)
+	if err != nil {
+		return nil, 0, errors.Wrapf(err, "create snapshot directory")
+	}
+
+	snapshot, from, immutableFiles, err := l.shouldSnapshot()
+	if err != nil {
+		return nil, 0, err
+	}
+	if !load && len(immutableFiles) == 0 {
+		// no snapshot needed and no need to load the state from disk
+		return nil, from, nil
+	}
+
+	// load the last snapshot
+	var state *DeserializationResult
+	if snapshot != "" {
+		start := time.Now()
+
+		l.logger.WithField("action", "hnsw_load_snapshot").
+			Info("loading snapshot")
+
+		state = readLastSnapshot(l.rootPath, l.id, l.logger)
+
+		if state != nil {
+			l.logger.WithField("action", "hnsw_load_snapshot").
+				WithField("duration", time.Since(start).String()).
+				Info("snapshot loaded")
+		}
+	}
+
+	if len(immutableFiles) == 0 {
+		// no commit log files to load, just return the snapshot state
+		if state == nil {
```

Comment:
is this conditional needed?

Reply:
Good question: yes it's needed because we need to specifically return 0 for `from` when the snapshot is corrupted.
I added a comment